### ISSUE TITLE
Reconsidering Windows Static FreeTDS Libraries

### ISSUE THREAD

**Issue by rsyring** [pymssql/pymssql#399]
I'd like to reconsider the decision to not ship FreeTDS statically in the windows binaries.  One of the primary goals that I have always had for pymssql was that it would be easily installable for those who are least likely to be able to build things from scratch.  That is why the binaries have always been statically linked.  I think asking our Windows users to build OpenSSL and FreeTDS is not the way to go.

I'm sorry that I'm just now realizing this was the direction that was taken and I apologize if I've missed discussions surrounding it.

My understanding based on the docs:

http://pymssql.org/en/v2.1.2/freetds.html#windows

Is that we had a problem that looked something like this:

- pymssql 2.1.1 was shipped without SSL support causing problems for those that wanted to connect to Azure
- Ramiro wanted to fix that problem, but realized that to fix it he would need to statically link in OpenSSL.  That presents a security problem since OpenSSL has many security vulnerabilities and should really be updated frequently.  This puts a burden on pymssql maintainers, of which, there r
really is only one right now.
- pymssql 2.1.2 therefore shipped without the static FreeTDS library.

**Short Term Fix Needed**

I believe that a short term fix for this problem is needed since this is a pretty big BC break and should not have been done in a point release.

Keep SSL out, put FreeTDS back in, and re-relase.  The problem to connect to Azure is still there, like it used to be, and we can revisit that in a different issue.

See #400.

**Security Implications of Shipping OpenSSL Statically**

A few thoughts:

- While I agree in principle with @ramiro's reason for not wanting to ship OpenSSL statically, I wonder if the implications are really that severe.  Do we believe that most people who build OpenSSL on Windows are going to be the type of people who will rebuild every time OpenSSL is released?
- Is it possible to ship a static version or give more advanced users a dynamic version?  I've seen something recently where an env variable was passed to pip which made a change, I just can't remember where.  Something like `PYMSSQL_DYNAMIC_FREETDS=true pip install pymssql`.
- What do other libraries do that use OpenSSL on Windows?
- Will the build pipeline that @ramiro has built lately make it easier to ship releases if OpenSSL has a serious vulnerability?

At the end of the day, and IMO, not shipping the libraries statically for Windows shouldn't really be an option for us.

**Thanks to Ramiro**

None of this is meant to take away anything from Ramiro or the work he has put in lately.  I'm grateful to see how much he has been getting done and thankful he is moving us along.

**Comment by rsyring** [comment]
From Ramiro on the ML regarding this issue:


For 2.1.2, I made the decision to stop shipping FreeTDS linked in staticall in pymssql and to depend on external FreeTDS build as DLLs instead.

In our docs (http://pymssql.org/en/v2.1.2/freetds.html#windows) we link to binaries of FreeTDS DLLS I maintain (https://github.com/ramiro/freetds/releases).

But there is more: These FreeDTS DLLs also depend on other DLLs: The OpenSSL ones chosen are these maintained at: http://www.npcglib.org/~stathis/blog/precompiled-openssl/.

(I decided to go with these because it's a binary distribution of OpenSSL which includes binaries built with the full gamut of Visual C versions used to build the official Windwos binaries of our supported Python versions [VC of VS 2008 for Python 2.7, VC of VS2010 for Python 3.3 and 3.4 and VC of VS2015 for Python 3.5] and thus won't risk causing strange errors at runtime because of mixed C runtime DLLs. This is also true for the FreeTDS DLLs above).

So, the situacion on Windows has changed like this:


```
              pymlssql 2.1.1                                           pymssql 2.1.2
+------------------------------------------+             +-------------------------------------------+
|                                          |             |                                           |
| pymssql                                  |             | pymssql                                   |
|                                          |             |                                           |
|                                          |             |                                           |
|                                          |  +------->  |                                           |
|                                          |             +-------------------------------------------+
|  +-----------------------------------+   |
|  | freetds (statically linked)       |   |             +-------------------------------------------+
|  |                                   |   |             |                                           |
|  |                                   |   |             | freetds (in DLL form)                     |
|  +-----------------------------------+   |             |                                           |
+------------------------------------------+             +-------------------------------------------+

                                                         +-------------------------------------------+
                                                         |                                           |
                                                         | OpenSSL (in DLL form)                     |
                                                         |                                           |
                                                         +-------------------------------------------+
```


Unfortunately, as the Python package tooling and file format don't allow to ship third party binaries, these DLLs can't be bundled with the pymssql Wheel binary file. And it woldn't be a good idea anyway (being the package responsible of putting extraneous DLL files on the global site-packages dir of the users' Python installation woldn't be something t be proud of.)

Why this? Because SSL is needed to open secure connections to SQL Server. Particularly it's required when one is trying to use pymssql to connect to Azure 'cloud' SQL Server flavor. And I didn't want to include OpenSSL linked statically with the pymssql binaries because we have a somewhat slow release cycle and in case security adivsores/udpates are published for OpenSSL I didn't want to either a) expose users to keep running older, insecure  binaries or b) Have to rush to a new release of pymssql simply because a new version of OpenSSL was released to fix security problems.

I'm still not sure this was a good deicision, and we might revisit it in future releases.

An alternative (which has been kown and used by many in prior releases) is to use pymssql binaries maintained (among many other) by Christoph Gohlke at http://www.lfd.uci.edu/~gohlke/pythonlibs/#pymssql

I hope this helps clifying the situation and the reasoning I followed when making these technical decisiones.

Regards,

**Comment by ramiro** [comment]
I'd like to say that IMHO this change in how the binaries for one platform are shipped is a backward-incompatible change but rather a lesser important change.

But I'm not opposed to shipping a 2.1.3 release as @rsyring proposes if anyone want to push it.

The releae process is automatized now, but it uses the DLLs so its possible things (build, and upload to PyPI) will need to be done manualy.

Fortunately the FreeTDS binaries I maintain and which are linked from the docs do containg static library versions so this make things easier because no local FreeTDS build needs to be performed.

I suspect, given the real backward-incompatible changes I perfomed on the master branch since the release (on our way to 2.2.0) , that a branch will need to be created from the point there 2.1.2 was tagged and the 2.1.3 release cut from that. I don't know if git/github do suport pointing nore than one tag to one unique commit.

**Comment by rsyring** [comment]
> rather a lesser important change.

Just for the sake of discussion, I agree that it's lesser important.  From a coding/dev perspective, it's not that big of a change.  However, from a user's point of view, `pip install pymssql` used to work without further intervention.  Now, it does not.  That is really important.

> But I'm not opposed to shipping a 2.1.3 release as @rsyring proposes if anyone want to push it.
> The releae process is automatized now, but it uses the DLLs so its possible things (build, and upload to PyPI) will need to be done manualy.

I'd be willing, but I don't really know where to begin.  Do we have docs for developers on what we need to do for a release?  The fact that you say it's both automated but will need to have some manual intervention confuses me a bit.

**Comment by InbarRose** [comment]
If I can weigh in on this, I use this module frequently. 

And my most common use is simple mssql, without any SSL requirement, from both windows and linux.

However, it should be possible, and could be nice to have _both_ the static libraries shipped with pymssql, and also the ability to specify an external one. Could be as simple as a parameter when initiating the connector. That way those who want to build their own libraries and maintain their own FreeTDS/OpenSSL can do so, and those who want a simple lightweight portable mssql solution in python also have that.

As it is, I am just not using the new version, and that shouldn't be the way a package is used.

Hope I am only adding, and not bothering.

**Comment by ramiro** [comment]
> I'd be willing, but I don't really know where to begin. Do we have docs for developers on what we need to do for a release? The fact that you say it's both automated but will need to have some manual intervention confuses me a bit.

One of the changes I introduced for 2.1.2 was the fact that the official release files published at PyPI are uploaded automatically from TravisCI and Appveyor when a git tag is created and pushed to GitHub after the corresponding CI run finishes successfully.

This helps a precisely in the case of the Windows wheels (which as per you proposal we agreed are the only binaries we publish) because it means the team member in charge of the relase doesn't need to have:
* A Windows system
* Installation of all the Python 2.x, 3.x versions we support both in x86 anf x86_64 flavors
* The thre MS C compilers need to buld binaries fully compatible with the officil python.org Python binaries/runtimes

But in the case of Windows+Appveyor the build links in the FreeTDS, OpenSSL libs dynamically:  https://github.com/pymssql/pymssql/blob/69850acb7bf52fe5d07a86d632e33fd889d54968/appveyor.yml#L161-L172 

If one wants to perform a release linking FreeTDS statically then I believe it would mean performing manually  what we  describe at:

https://github.com/pymssql/pymssql/blob/master/README_building_and_developing.rst#ms-visual-c 

for every cell in the python+C compiler versions x platform matrix mentioned above. In fact that's what I remember doing when I released 2.1.1 (had temporary access to a Windows system to play with and break).

Another option would be to  modify the AppVeyor script.

Unfortunately PEP 427 (https://www.python.org/dev/peps/pep-0427/#file-name-convention) doesn't provide i.e. `pymssql-static_freetds-2.1.3....whl` plus `pymssql-dynamic_freetds-2.1.3....whl`. Maybe http://pythonhosted.org/setuptools/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies can be of help here? e.g. `pip install pymssql[ssl]`to install a version that depends on the extra DLLs?

**Comment by lsuttle** [comment]
Feedback from the sort of user you were referring to (not easily able to build from scratch): I have used this for over a year for my work and really appreciate its availability. I updated today and have spent hours trying to fix this install with no success. I'm going to have to move on to another package due to this change unless it is reverted, as I rely on this for one of my data pipelines.

**Comment by rsyring** [comment]
@lsuttle you have some other options:

-     Follow the instructions here: http://pymssql.org/en/v2.1.2/freetds.html#windows
-     Download unofficial but recommended binaries: http://www.lfd.uci.edu/~gohlke/pythonlibs/#pymssql
-     Install the previous version of pymssql: https://pypi.python.org/pypi/pymssql/2.1.1

We'll try to get this sorted out in the next few weeks.

**Comment by lsuttle** [comment]
The second option worked for me! Thank you, I really wasn't looking forward to tomorrow and remaking that pipeline.

**Comment by pedrobraz1990** [comment]
Hello everyone. I'm glad to see how much effort the team (Randy and Ramiro) are putting on it. Thanks for that guys, and congratulations on your work.

I tried sending an email on the mailing list but I'll do it here as well.

I'm more interested in the third option, installing older versions with ``pip``.

I've tried the commands:

    pip install pymssql==2.1.1
    pip install pymssql==2.1.0
    pip install pymssql==2.0.1

none of them works. How did you guys manage to do it?

**Comment by DavidRueter** [comment]
While I like the simplicity of `pip install pymssql`, in order to be able to fix problems as they arise I really need to be able to build everything from source.

I'm running Python 3.4 on Windows.  I had no problem doing this with dynamic freetds + openssl (other than the learning curve and hassle of setting up to build everything).  Plus, by building these three packages separately and linking dynamically, I can update each as needed.

I have been unable to get pymssql to build using static freetds with no openssl.  I'm sure I could eventually figure it out, but so far I've had no luck due to errors when building in this way.

I may be unusual, but I prefer dynamic freetds + openssl, prefer to be able to rebuild everything as needed, and prefer to use my own binaries.  That said, as long as switching to this remains as simple as it currently is (updating two constant values in setup.py), I guess it doesn't matter too much to me how the default binaries are set up.  (I think that not being able to use Azure would be a problem for some folks though.)

I'd be in favor of seeing the GitHub wiki be used to document build procedures and various other tips.  Is there a reason this project doesn't make use of the wiki?

**Comment by rsyring** [comment]
@DavidRueter you can build it from source.  If you need instructions, there may be information in the Appveryor config that will help you:

https://github.com/pymssql/pymssql/blob/master/appveyor.yml

> I'd be in favor of seeing the GitHub wiki be used to document build procedures and various other tips. Is there a reason this project doesn't make use of the wiki?

We use Sphinx for documentation.  Check out the docs directory in the source or http://pymssql.org/en/latest/ for the published version.

It would be great if our docs would be better.  But right now, we really only have Ramiro working on the project.  It's just a time/labor issue.

**Comment by msabramo** [comment]
I did some builds on Windows a long time ago and back then it was painful, I think because at the time FreeTDS needed all kinds of hacks. Hopefully that's better now.

Like @rsyring said, pretty much only @ramiro has been working on the project, so it's a challenge to get a lot of things done.

If you manage to figure out how to build on Windows, it would be awesome if you sent a PR that adds instructions to the Sphinx docs. That would probably make a lot of people happy including myself and your future self.

**Comment by DavidRueter** [comment]
I am able to build OpenSSL, freetds, and pymssql on Windows 10 / Python 3.4 / Visual Studio 2010, and I am happy to document or help in any way I am able.

However, at this point I can't make my build of freetds work with pymssql using static linking.  My build of freetds seems to works with pymssql with dynamic linking just fine.  However my build of pymssql works with static linking if I use the binary of freetds provided by @ramiro 

In other words, I'd love to figure out how to get my build of freetds to work with static linking in the same way that the freetds binaries from @ramiro work.  Then I'd feel pretty comfortable documenting.

Do we have any information on how the builds of freetds on https://github.com/ramiro/freetds/releases were produced (that use win-iconv)?

**Comment by ramiro** [comment]
@DavidRueter : We already have some pointers in our docs:  http://pymssql.org/en/latest/freetds.html#windows

In particular, that Windows specific section links to https://github.com/ramiro/freetds/releases

There, you can take a look at the https://github.com/ramiro/freetds/blob/freetds-0.95-for-pymssql/misc/appveyor.yml file as the FreeTDS static libs we use are built on AppVeyor. It's well commented so it should give you some ideas bout how to bulld and/or integrate all the dependencies. The AppvEyor setup is a variation (specifically tailored so its more easily usable to build pymssql) of the Windows CI  setup I contributed to FreeTDS. 

+1 to to have any documentation contribution added to our docset instead of the GitHub  Wiki.

**Comment by ramiro** [comment]
This issue was 'fixed' in v2.1.3.
